test/codec-test.hs

```diff
@@ -0,0 +1,271 @@
+
+import           Control.Monad (when)
+import           Control.Monad.ST.Strict (ST, runST)
+import           Data.Foldable (fold)
+import           Data.STRef
+import           Data.Vector (Vector)
+import qualified Data.Vector as V
+import           Database.LSMTree.Common (BloomFilterAlloc (..),
+                     DiskCachePolicy (..), NumEntries (..), TableConfig (..),
+                     WriteBufferAlloc (..), defaultTableConfig)
+import           Database.LSMTree.Internal.MergeSchedule (NumRuns (..))
+import           Database.LSMTree.Internal.RunNumber (RunNumber (..))
+import           Database.LSMTree.Internal.Snapshot
+import           Database.LSMTree.Internal.Snapshot.Codec
+import qualified System.FS.API as FS
+import           System.FS.API.Types (FsPath, MountPoint (..), fsToFilePath,
+                     mkFsPath, (<.>))
+import           System.FS.IO (HandleIO, ioHasFS)
+import qualified Test.Tasty as Tasty
+import           Test.Tasty (TestName, TestTree, testGroup)
+import qualified Test.Tasty.Golden as Au
+import           Text.Printf (printf)
+
+
+-- |
+-- Delete output files on test-case success.
+-- Change the option here if this is undesireable.
+handleOutputFiles :: TestTree -> TestTree
+handleOutputFiles = Tasty.localOption Au.OnPass
+
+-- |
+--
+-- Internally, the function will infer the correct filepath names.
+--
+-- There are three paths for both the checksum and the snapshot files:
+--   1. The filepath of type @FsPath@ to which data is written.
+--   2. The filepath of type @FilePath@ from which data is read.
+--   3. The filepath of type @FilePath@ against which the data is compared.
+--
+-- These file types' bindings have the following infix annotations, respectively:
+--   1. (Fs) for FsPath
+--   2. (Hs) for "Haskell" path
+--   3. (Au) for "Golden file" path
+snapshotCodecTest
+  :: String -- ^ Name of the test
+  -> SnapshotMetaData -- ^ Data to be serialized
+  -> TestTree
+snapshotCodecTest name datum =
+  let -- Various paths
+      checksumFsPath = mkFsPath [name] <.> "checksum"
+      checksumHsPath = fsToFilePath mountingAtPath checksumFsPath
+      checksumAuPath = checksumHsPath <> ".golden"
+      snapshotFsPath = mkFsPath [name] <.> "snapshot"
+      snapshotHsPath = fsToFilePath mountingAtPath snapshotFsPath
+      snapshotAuPath = snapshotHsPath <> ".golden"
+      mountingAtPath = MountPoint "test/codec-data"
+
+      -- IO actions (or lack thereof)
+      noWorkAction :: IO ()
+      noWorkAction = pure ()
+      tidyUpAction :: a -> IO ()
+      tidyUpAction = const noWorkAction
+      runnerIO :: FS.HasFS IO HandleIO
+      runnerIO = ioHasFS mountingAtPath
+      removeIfExists :: FsPath -> IO ()
+      removeIfExists fp =
+        FS.doesFileExist runnerIO fp >>= (`when` (FS.removeFile runnerIO fp))
+      outputAction :: IO ()
+      outputAction = do
+        removeIfExists snapshotFsPath
+        removeIfExists checksumFsPath
+        writeFileSnapshotMetaData
```

Comment:
I think we should use [`toLazyByteString`](https://hackage.haskell.org/package/cborg-0.2.10.0/docs/Codec-CBOR-Write.html#v:toLazyByteString) on the result of [`encode`](https://github.com/IntersectMBO/lsm-tree/blob/main/src/Database/LSMTree/Internal/Snapshot/Codec.hs#L136-L137) and write it to disk. We can leave out the checksums

Reply:
Reworked the serialization methodology.